### PR TITLE
Implementation of rule 2A

### PR DESCRIPTION
**Pull request by OisinNolan**
For reference, rule 2A:
`If the current node is hidden and is not directly referenced by aria-labelledby or aria-describedby, nor directly referenced by a native host language text alternative element (e.g. label in HTML) or attribute, return the empty string.`

- Added gts package (https://www.npmjs.com/package/gts) to check for style errors.
- Added file src/lib/rules.ts to store & label the functions implementing each of the rules (2.A - 2.I).
- These rule functions are called sequentially by _getAccessibleName() in src/accessible_name.ts
- Each rule function returns a result string if that rule may be applied (according to its condition), or null otherwise.
- Added Context interface in src/lib/context.ts to keep track of relevant additional information for a given element as it's passed into a rule function.
- 'hidden_node.ts' implements rule 2A. The hiddenNodeCondition() helper function checks whether or not rule 2A may be applied to the element whose accessible name is being computed.
- Tests for hiddenNode in 'src/lib/hidden_node_tests.ts' - (May need to add more)

In this implementation I've attempted to write a system that will allow me to add new rules easily. Adding a new rule would involve implementing a rule function that returns a string if that rule may be applied to the given element, or null otherwise. Once the function is written, all that needs to be done is to add the function to the rules map in 'src/lib/rules.ts'.

### CODE REVIEW

File: src/accessible_name.ts [hunk 1]
```diff
@@ -1,12 +1,32 @@
...
+/**
+ * @param {HTMLElement} elem - The element whose accessible name will be calculated
+ * @param {Context} context - Additional information relevant to the name computation for elem
+ * @return {string} - The accessible name for elem
  */
-export function getTextNodeAccessibleName(textNode: Node): string {
-    let accessibleName: string = '';
-    if (textNode && textNode.nodeType === Node.TEXT_NODE && textNode.textContent) {
-        accessibleName = textNode.textContent;
+function _getAccessibleName(elem: HTMLElement, context: Context): string {
```
Comment (engelsdamien) [inline src/accessible_name.ts:9]: don't use '_' as a prefix. Prefer naming your function with an appropriate name (computeTextAlternative might be good)

also getAccessibleName can keep an HTMLElement as a param, but this should probably take a Node.

File: src/accessible_name.ts [hunk 2]
```diff
@@ -1,12 +1,32 @@
...
+  let accessibleName = '';
+  let result;
+  // Iterate over the rules used to calculate the accessible name, stop
+  // iterating if a valid string is returned (i.e. if result is not null).
+  const ruleKeys = Object.keys(rules);
```
Comment (engelsdamien) [inline src/accessible_name.ts:14]: While I understand the value of modularizing the code in different functions for each rule, I think the main function should still refer to each step explicitly. Otherwise the code is much harder to read.

File: src/lib/context.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/**
+ * This interface will be used to pass additional information
+ * about the element whose name is being computed.
+ */
+export interface Context {
+  ariaLabelledbyReference?: boolean;
+  nativeTextAlternativeReference?: boolean;
```
Comment (engelsdamien) [inline src/lib/context.ts:7]: can you add some documentation on what each of those flags mean ?

File: src/lib/hidden_node.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import {Context} from './context';
+
+/**
+ * Condition for applying rule 2A
+ * @param {HTMLElement} elem - The element whose accessible name is being calculated
+ * @param {Context} context - Additional information relevant to the name computation for elem
+ * @return {boolean} - Whether of not elem satisfies the condition for rule 2A
+ */
+function hiddenNodeCondition(elem: HTMLElement, context: Context): boolean {
+  const hidden: boolean = elem.hasAttribute('hidden');
```
Comment (engelsdamien) [inline src/lib/hidden_node.ts:10]: 'hidden' is actually not well defined in the acc-name document, but I am pretty sure it is not just based on the 'hidden' attribute.

I think you will also need to check aria-hidden="true" & potentially CSS-based ways of hiding things. Typically having display: none also causes a node to be considered "hidden". I am not sure if there is a place that defines this precisely.

Comment (engelsdamien) [reply src/lib/hidden_node.ts:10]: also note that hidden applies to an entire subtree, so the node is considered hidden if any of its parents are hidden.

File: src/lib/hidden_node.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import {Context} from './context';
+
+/**
+ * Condition for applying rule 2A
+ * @param {HTMLElement} elem - The element whose accessible name is being calculated
+ * @param {Context} context - Additional information relevant to the name computation for elem
+ * @return {boolean} - Whether of not elem satisfies the condition for rule 2A
+ */
+function hiddenNodeCondition(elem: HTMLElement, context: Context): boolean {
```
Comment (engelsdamien) [inline src/lib/hidden_node.ts:9]: I don't think that "hiddenNodeCondition" is a very descriptive name here. This file seems to abstract step 2A, so we should probably name it "step2ACondition" or something similar. 

The fact that a node is "hidden" is a separate concern that is only part of step 2A. Even if a "hidden" node is referenced by aria-labelledby, it should still be considered "hidden".

File: src/lib/hidden_node_test.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  it('returns empty string for hidden elems that aren\'t referenced', () => {
+    render(html`<div id="foo" hidden>Hello world</div>`, container);
+    const elem = document.getElementById('foo');
```
Comment (engelsdamien) [inline src/lib/hidden_node_test.ts:18]: you should also be able to say `container.firstElementChild`

File: src/lib/hidden_node_test.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    if (elem) {
```
Comment (engelsdamien) [inline src/lib/hidden_node_test.ts:19]: instead of checking if the `elem` is null, you can also just assert it since you know that it cannot be null:

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

File: tslint.json [hunk 8]
```diff
@@ -1,7 +1,6 @@
  
 {
   "rules": {
-    "arrow-parens": true,
```
Comment (engelsdamien) [inline tslint.json:4]: why did you need to remove this ?

Comment (OisinNolan) [reply tslint.json:4]: I think it was conflicting with another one of the style guides (probably gts)

File: src/accessible_name.ts (hunk 2 above)
Comment (OisinNolan) [reply src/accessible_name.ts:14]: I agree, I've added some more suggestions to the doc.

File: src/lib/hidden_node_test.ts (hunk 7 above)
Comment (OisinNolan) [reply src/lib/hidden_node_test.ts:19]: Really useful, thanks!

File: src/lib/rule2A.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      node.hasAttribute('hidden') ||
+      node.getAttribute('aria-hidden') === 'true' ||
+      node.offsetHeight <= 0 ||
+      node.offsetWidth <= 0 ||
+      hiddenByStyle
+    );
+  }
+  // A node is considered hidden if any ancestor is hidden
+  const parentNode = node.parentNode;
+  if (parentNode) {
```
[the remark this answers is not in the document]
Comment (engelsdamien) [reply src/lib/rule2A.ts:28]: I think the best way to do this efficiently is probably to use `getComputedStyles` for css based hiding (automatically checks ancestors)

and then for the other properties you can use Element.closest to quickly find if there is a hidden ancestor (it also includes the node itself):

`node.closest('[hidden],[aria-hidden="true"]')` should return null if there is no hidden ancestor

File: src/lib/rule2A_test.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
+import {rule2A} from './rule2A';
+import {html, render} from 'lit-html';
+
+describe('The hiddenNode function for rule 2A', () => {
```
Comment (engelsdamien) [inline src/lib/rule2A_test.ts:4]: update

File: src/lib/rule2A.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import {Context} from './context';
+
+/**
+ * Looks at a variety of characteristics (CSS, size on screen, attributes) 
+ * to determine if 'node' should be considered hidden
```
[the remark this answers is not in the document]
Comment (OisinNolan) [reply src/lib/rule2A.ts:5]: I didn't actually find an authoritative definition for this but used JQuery's definition for visible (offsetHeight / offsetWidth) combined with some techniques for hiding elements with CSS. There may be more ways to consider than this but I think the offsetHeight/Width handle any case where the element doesn't take up screen space (including display:none, as far as I can tell)

[the remark this answers is not in the document]
Comment (engelsdamien) [reply src/lib/rule2A.ts:5]: let's stick to a simpler definition for now then as this is the only one present in some spec:
 - hidden
 - aria-hidden=true
 - style="display: none;"
 - style="visibility: hidden;" 

also include the link above in the function's jsdoc so that we know where the implementation is derived from.

we can always make it more complex later if/when we see cases where the browser disagrees. But we'll probably also want to also file a bug against the spec.

File: src/accessible_name.ts [hunk 12]
```diff
@@ -1,12 +1,30 @@
...
+  return '';
+}
+
+/**
+ * Main exported function for the library. Initialises traversal with an empty context.
+ * @param elem - The element whose accessible name will be calculated
+ * @return - The accessible name for elem
+ */
+export function getAccessibleName(elem: HTMLElement): string {
+  const initialContext = {};
```
Comment (engelsdamien) [inline src/accessible_name.ts:28]: nit: inline initialContext.

File: src/lib/rule2A.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+      // measurement of pixels used on screen (checks for display:none)
+      node.offsetHeight <= 0 ||
+      node.offsetWidth <= 0
+    );
+
+    const hiddenByAttribute = (
+      node.closest('[hidden],[aria-hidden="true"]') !== null
+    );
+
+    return (hiddenByAttribute || hiddenByStyle);
```
Comment (engelsdamien) [inline src/lib/rule2A.ts:24]: style nit: you are potentially computing unnecessary things here by not returning early. 

Returning early can also help with readability & avoid levels of nesting.

I'd suggest organizing the code like the following:

```typescript
function isHidden(node: Node): boolean {
  if (!(node instanceof HTMLElement)) {
    return false;
  }

  const visibility = window.getComputedStyle(node).visibility;
  const notDisplayed = node.offsetHeight === 0 && node.offsetWidth === 0;
  if (visibility === 'hidden' || notDisplayed) {
    return true;
  }

  const hiddenAncestor = node.closest('[hidden],[aria-hidden="true"]');
  if (hiddenAncestor !== null) {
    return true;
  }

  return false;
}
```

File: src/lib/rule2A.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+ */
+function isHidden(node: Node): boolean {
+  if (node instanceof HTMLElement) {
+    const nodeStyle = window.getComputedStyle(node);
+
+    const hiddenByStyle = (
+      nodeStyle.visibility === 'hidden' ||
+      // measurement of pixels used on screen (checks for display:none)
+      node.offsetHeight <= 0 ||
+      node.offsetWidth <= 0
```
Comment (engelsdamien) [inline src/lib/rule2A.ts:17]: I think we want an `&&` between those two conditions to check for `display: none`

File: src/lib/rule2A.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+ * @return - whether or not the node is considered hidden
+ */
+function isHidden(node: Node): boolean {
+  if (node instanceof HTMLElement) {
+    const nodeStyle = window.getComputedStyle(node);
+
+    const hiddenByStyle = (
+      nodeStyle.visibility === 'hidden' ||
+      // measurement of pixels used on screen (checks for display:none)
+      node.offsetHeight <= 0 ||
```
Comment (engelsdamien) [inline src/lib/rule2A.ts:16]: Can this be `< 0` ?

If not, change to `===`

File: src/lib/rule2A_test.ts [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    expect(rule2A(elem!, {labelReference: true})).toBe(null);
+  });
+
+  it('returns null for elem that is not hidden', () => {
+    render(html`<div id="foo">Hello world</div>`, container);
+    const elem = document.getElementById('foo');
+    expect(rule2A(elem!, {})).toBe(null);
+  });
+
+  it('considers aria-hidden', () => {
```
Comment (engelsdamien) [inline src/lib/rule2A_test.ts:40]: starting from this test, all the remaining cases are testing `isHidden` directly.

you can use an export named TEST_ONLY to export the function for the test

```typescript
export const TEST_ONLY = {isHidden};
```

Comment (OisinNolan) [reply src/lib/rule2A_test.ts:40]: I'm not sure what you mean here, would I test isHidden directly in the _test.ts file referring to it as TEST_ONLY?

When I try:
```
it('considers aria-hidden', () => {
    render(html`<div id="foo" aria-hidden="true">Hello world</div>`, container);
    const elem = document.getElementById('foo');
    expect(TEST_ONLY(elem!)).toBe(true);
  });
```
I get an error saying that TEST_ONLY is not callable.
